apps/aechannel/src/aesc_fsm.erl

```diff
@@ -2899,54 +3074,103 @@ check_tx_and_verify_signatures(SignedTx, Updates, Mod, Data, Pubkeys, ErrTypeMsg
-            case Mod:channel_pubkey(Tx) of
-                ChannelPubkey -> %% expected pubkey
-                    CorrectRound =
-                        case Mod of
-                            aesc_close_mutual_tx -> true; %% no round here
-                            _ -> Mod:round(Tx) =:= ExpectedRound
-                        end,
-                    case CorrectRound of
-                        false ->
-                            {error, wrong_round};
-                        true when is_list(Updates) ->
-                            case check_update_tx(SignedTx, Updates, State,
-                                                 Opts, ChannelPubkey) of
-                                ok ->
-                                    case Mod of
-                                        aesc_offchain_tx ->
-                                            verify_signatures_offchain(ChannelPubkey,
-                                                                       Pubkeys,
-                                                                       SignedTx);
-                                        _ ->
-                                            verify_signatures_onchain_check(Pubkeys,
-                                                                            SignedTx)
-                                    end;
-                                {error, E} -> {error, E}
-                            end
-                    end;
-                _ ->
-                    {error, different_channel_id}
-            end;
+            Checks =
+                [ fun() -> check_channel_id(Mod, Tx, ChannelPubkey) end,
+                  fun() -> check_round(Mod, Tx, ExpectedRound) end,
+                  fun() ->
+                      check_update_tx(SignedTx, Updates, BlockHash,
+                                      State, Opts, ChannelPubkey)
+                  end
+                ],
+            aeu_validation:run(Checks);
         E ->
             lager:debug("_E = ~p", [E]),
             {error, ErrTypeMsg}
     end.
 
+-spec check_tx_if_expected(aetx_sign:signed_tx(), aetx_sign:signed_tx(),
+                           atom()) -> ok | {error, atom()}.
+check_tx_if_expected(NewSignedTx, OldSignedTx, WrongMod) ->
+    {Mod, Tx} =
+        aetx:specialize_callback(aetx_sign:innermost_tx(NewSignedTx)),
+    {OldMod, OldTx} =
+        aetx:specialize_callback(aetx_sign:innermost_tx(OldSignedTx)),
+    ExpectedRound = get_round(OldMod, OldTx),
+    ExpectedHash = get_state_hash(OldMod, OldTx),
+    Checks =
+        [ fun() -> check_channel_id(Mod, Tx, OldMod:channel_pubkey(OldTx)) end,
+          fun() -> check_round(Mod, Tx, ExpectedRound) end,
+          fun() -> check_state_hash(Mod, Tx, ExpectedHash) end,
+          fun() ->
+              case Mod =:= OldMod of
+                  true -> ok;
+                  false -> {error, WrongMod}
+              end
+          end
+        ],
+    aeu_validation:run(Checks).
+
+get_round(Mod, Tx) ->
```

Comment:
I'd prefer function clauses here instead of the use of a case clause, same for the functions coming next.